pkgs/applications/misc/dmenu-rs/default.nix

```diff
@@ -0,0 +1,84 @@
+{ stdenv
+, rustPlatform
+, lib
+, cargo
+, clang
+, expat
+, fetchFromGitHub
+, fontconfig
+, libXft
+, libXinerama
+, m4
+, pkg-config
+, python3
+}:
+
+# The dmenu-rs package has extensive plugin support. However, this derivation
+# only builds the package with the default set of plugins. If you'd like to
+# further customize dmenu-rs you can build it from the source.
+# See: https://github.com/Shizcow/dmenu-rs#plugins
+stdenv.mkDerivation rec {
+  pname = "dmenu-rs";
+  version = "5.5.1";
+
+  src = fetchFromGitHub {
+    owner = "Shizcow";
+    repo = pname;
+    rev = version;
+    sha256 = "sha256-WpDqBjIZ5ESnoRtWZmvm+gNTLKqxL4IibRVCj0yRIFM=";
+  };
+
+  nativeBuildInputs = [
+    cargo
+    clang
+    m4
+    pkg-config
+    python3
+    rustPlatform.bindgenHook
+    rustPlatform.cargoSetupHook
+  ];
+
+  buildInputs = [
+    expat
+    fontconfig
+    libXft
+    libXinerama
+  ];
+
+  # The dmenu-rs repository does not include a Cargo.lock because of its
+  # dynamic build and plugin support. Generating it with make and checking it
+  # in to nixpkgs here was the easiest way to supply it to buildRustPackage.
+  # See: https://github.com/Shizcow/dmenu-rs/issues/34#issuecomment-757415584
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+  };
+
+  # The rust-xcb dependency dynamically generates rust code at build time.
+  # This derivation uses nixpkgs rust functions that vendor each cargo
+  # dependency's source code into the READ-ONLY nix store. To avoid the code
+  # generation step failing, we copy the rust-xcb source out of the nix store
+  # and make it writeable.
+  # See: https://github.com/rust-x-bindings/rust-xcb/tree/v0.8.2
+  patchPhase = ''
+    chmod +w "$NIX_BUILD_TOP/cargo-vendor-dir"
+    mkdir -p "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    cp -r --no-preserve=mod "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2/." "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite"
+    unlink "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+    mv "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2-readwrite" "$NIX_BUILD_TOP/cargo-vendor-dir/xcb-0.8.2"
+  '';
+
+  preConfigure = ''
+    sed -i "s@PREFIX = /usr/local@PREFIX = $out@g" config.mk
+  '';
+
+  # Running make test requires an X11 server, so we skip it.
+  doCheck = false;
```

Comment:
you can probably run tests with xvfb-run

Reply:
I've updated the `checkPhase` to use the following

```bash
timeout 30s xvfb-run make test || if [ "$?" == "124" ]; then true; else false; fi
```

but ... personally, I would prefer no check phase at all.

This test only passes if the 30 seconds is long enough for the test to run. It seemed to work pretty consistently with 5 seconds on my machine, but using a timeout like this seems destined to cause flaky failures somewhere.

The reason I reached for a timeout was because the `make test` hangs on user input, and I don't how to supply any input keys to the program or avoid this hanging test.

For more detail see: https://github.com/NixOS/nixpkgs/pull/200977#discussion_r1027367488

Reply:
My bad I had no read the full discussion. If the tests are not well behaved enough to terminate, let's forget about them.

Reply:
Cool, dropped the test.